sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala

```diff
@@ -0,0 +1,165 @@
+    getDataFrame(generateSeqInput(l).map(_.toString))
+  }
+
+  def benchmarkUTFStringEquals(collationTypes: Seq[String], utf8Strings: Seq[UTF8String]): Unit = {
+    val sublistStrings = utf8Strings
+
+    val benchmark = new Benchmark(
+      "collation unit benchmarks - equalsFunction",
+      utf8Strings.size * 10,
+      warmupTime = 4.seconds,
+      output = output)
+    collationTypes.foreach(collationType => {
+      val collation = CollationFactory.fetchCollation(collationType)
+      benchmark.addCase(s"$collationType") { _ =>
+        sublistStrings.foreach(s1 =>
+          utf8Strings.foreach(s =>
+            (0 to 10).foreach(_ =>
+              collation.equalsFunction(s, s1).booleanValue())
+          )
+        )
+      }
+    }
+    )
+    benchmark.run()
+  }
+  def benchmarkUTFStringCompare(collationTypes: Seq[String], utf8Strings: Seq[UTF8String]): Unit = {
+    val sublistStrings = utf8Strings
+
+    val benchmark = new Benchmark(
+      "collation unit benchmarks - compareFunction",
+      utf8Strings.size * 10,
+      warmupTime = 4.seconds,
+      output = output)
+    collationTypes.foreach(collationType => {
+      val collation = CollationFactory.fetchCollation(collationType)
+      benchmark.addCase(s"$collationType") { _ =>
+        sublistStrings.foreach(s1 =>
+          utf8Strings.foreach(s =>
+            (0 to 10).foreach(_ =>
+              collation.comparator.compare(s, s1)
+            )
+          )
+        )
+      }
+    }
+    )
+    benchmark.run()
+  }
+
+  def benchmarkUTFStringHashFunction(
+      collationTypes: Seq[String],
+      utf8Strings: Seq[UTF8String]): Unit = {
+    val sublistStrings = utf8Strings
+
+    val benchmark = new Benchmark(
+      "collation unit benchmarks - hashFunction",
+      utf8Strings.size * 10,
+      warmupTime = 4.seconds,
+      output = output)
+    collationTypes.foreach(collationType => {
+      val collation = CollationFactory.fetchCollation(collationType)
+      benchmark.addCase(s"$collationType") { _ =>
+        sublistStrings.foreach(_ =>
+          utf8Strings.foreach(s =>
+            (0 to 10).foreach(_ =>
+              collation.hashFunction.applyAsLong(s)
+            )
+          )
+        )
+      }
+    }
+    )
+    benchmark.run()
+  }
+
+  def benchmarkFilterEqual(
+      collationTypes: Seq[String],
+      dfUncollated: DataFrame): Unit = {
+    val benchmark =
+      new Benchmark(
+        "filter df column with collation",
+        dfUncollated.count(),
+        warmupTime = 4.seconds,
+        output = output)
+    collationTypes.foreach(collationType => {
+      val dfCollated = dfUncollated.selectExpr(
+        s"collate(s2, '$collationType') as k2_$collationType",
+        s"collate(s1, '$collationType') as k1_$collationType")
+      benchmark.addCase(s"filter df column with collation - $collationType") { _ =>
+        dfCollated.where(col(s"k1_$collationType") === col(s"k2_$collationType"))
```

Comment:
do we really need this benchmark? Isn't just the equalsFunction?

Reply:
It is a fair question. For one, it is good proof of concept that the performance differences at the internal unit level do translate to actual performance regressions when actual users use spark. Showing that improving the performance is actually going to be worth our time.

Reply:
We have `HashBenchmark` which does not test end-to-end queries either. I think it's fair to trust the system is reasonable and will translate expression level performance to end-to-end query performance.

Reply:
@cloud-fan  Okay. I defer to you and @dbatomic as both of you wanted it removed.

so I removed the e2e testing.